decisions/DR008-Combine-getEntityAttributes-and-resovle.md

```diff
@@ -0,0 +1,329 @@
+were authored outside of OpenAssetIO interactions, that would require
+work to determine what a suitable primary string (or string trait value)
+should be.
+
+We believe that in real-world scenarios, any overhead the data
+structures required to support the trait-based return types would be
+relatively insignificant compared to all other work performed by a
+manager to satisfy any given lookup.
+
+[TBC empirical measurements of the differences between the two
+approaches in terms of the low-level API mechanics, correlated against
+practical timings from real-world asset systems].
+
+## Appendix
+
+What follows in a comparative illustration of the two approaches, the
+code for Option 1 assumes that the work has been done to make attributes
+versionable and consistent with
+[#4](https://github.com/TheFoundryVisionmongers/OpenAssetIO/issues/4),
+and that the relevant API methods have been updated to work with trait
+sets.
+
+### Preamble
+
+The examples assume the following traits and specifications are
+available - allowing a file-based image to be described:
+
+```cpp
+trait BlobTrait {
+    name: "blob",
+    props: { url: str, mimeType: str }
+}
+
+trait ImageTrait {
+    name: "image",
+    props: { colorSpace: str }
+}
+
+spec ImageSpec {
+    traits: ["blob", "image"]
+}
+```
+
+### Publishing an image
+
+This example illustrates a host wanting to publish an image with its
+associated color space, to ensure it is accurately tracked.
+
+```python
+img = ImageSpec()
+img.blobTrait().setUrl("file:///some/render.exr")
+img.imageTrait().setColorSpace("ACES - ACES2065-1")
+```
+
+#### Option 1
+
+```python
+manager.register(
+    "file:///some/render.exr", img.toAttributes(),
+    img.traitNames, target_ref, context)
+```
+
+This is ambiguous, as should the blob trait properties be included in
+`toAttributes`? The `url` is the primary string, but `mimeType` needs
+to go via attributes (if it was set).
+
+#### Option 2
+
+```python
+manager.register(img, target_ref, context)
+```
+
+Removes the ambiguity as the spec, complete with its data for all
+applicable traits is registered in one go, with no duplication or need
+to work out 'which single property' becomes the primary string.
+
+### Resolving data for an image
+
+This example shows a host wanting to read an image with the correct
+color space.
+
+#### Option 1
+
+```python
+path = manager.resolveEntityReference(img_ref, context)
+attr = manager.getEntityAttributes(img_ref, ImageTrait.attrNames, context)
+
+color_space = ImageTrait(attr).getColorSpace()
+```
+
+Note the asymmetry to `register`, only the `ImageTrait` is needed for
+lookup as otherwise if `ImageSpecification` was used, it would also
+request the already accessed `url` from the blob, which we "know" we
+already have through the primary string. Maybe we should have omit
+resolve and just fetched it all via attributes?
+
+#### Option 2
+
+```python
+data = manager.resolve(img_ref, ImageSpec.traitNames, context)
```

Comment:
I think the proposal looks good overall. I was only wondering about corner cases where someone might want to do a bulk call for resolving many different refs (and only cares about the returned path), which may be cheaper if the manager doesn't have to create the full set of attributes for all of them. I expect though this is rare and the amount of data involved is small, so it's unlikely to be a problem.

Reply:
Cool, thanks.  That situation is definitely a concern. The closest you could get with this is the single trait use below, which then is limited to just the (available) properties of the `BlobTriat` which would be the path part. You could request this for any entity that has that trait.

```python
path = BlobTrait(manager.resolve(image_refs, [BlobTrait.name], context)).getUrl()
```

Reply:
Seems reasonable I think. The cost of communication inside the AMS will likely outweigh any of the OpenAssetIO code.